<a href="https://colab.research.google.com/github/String-ru/AIFFEL_quest_cr/blob/main/Datathon/1st_datathon_0220_6PM_1st_assumption.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# C1 DAtathon

# DOTT 팀

# 1st Assumption

#특정 장르에서 특정 배우는 더 높은 선호도를 보일까?

- 장르에 대한 배우 개인의 출연 선호도 X

- 여러 장르에 출연하는 배우에 대한 시청자의 선호도 O

#개요

# 목차 구성을 완료하면 목차 기반으로 개요 다시 정리할 예정

# 목차

# 1. Data 불러오기

Google Drive 마운트 및 OTT 데이터셋 로드, 컬럼 정리, 플랫폼 정보 추가 완료. ✅

# 2. 모든 dataset에서 column 동일한지 확인

모든 OTT 데이터셋의 컬럼이 동일한지 검증 완료. ✅

# 3. 결측치 확인

각 데이터셋의 결측치 개수 및 비율 확인 완료. ✅

# 4. 중복data 확인

중복 데이터 없음 확인 완료. ✅

# 5. 결측치 처리 및 data 병합

결측치 처리 및 데이터 병합 완료, 최종 데이터 19,925행 10컬럼 생성. ✅

# 6. 중복되는 영화/TV프로그램 제거

중복된 영화/TV 프로그램을 플랫폼 우선순위로 정리하여 중복 제거 완료, 최최종 데이터 8,804행 9컬럼. ✅

# 7. 장르 파악 (listed_in)

'listed_in'에서 장르를 분리하고 출현 빈도 계산 완료, 상위 20개 장르 출력. ✅

# 8. 전체 장르 고유값 파악

'listed_in'에서 고유한 장르 목록 추출 완료. ✅

# 9. 장르 개수 파악 및 대분류 카테고리화

'listed_in'에서 세부 장르를 분리하여 유니크한 장르 42개 추출 완료. ✅

# 10. 10개 대분류 카테고리와 해당 세부 장르 매핑

10개 대분류 카테고리와 세부 장르 매핑 완료. ✅

# 11. 대분류 카테고리화

listed_in 컬럼을 기반으로 세부 장르를 대분류 카테고리로 매핑하여 새로운 category 컬럼을 추가 ✅

# 12. 한 영화/TV당 한 장르 배정

각 작품에 대해 가장 우선순위가 높은 카테고리를 top_category 컬럼으로 배정 완료. ✅

# 1. Data 불러오기

Google Drive 마운트 및 OTT 데이터셋 로드, 컬럼 정리, 플랫폼 정보 추가 완료. ✅

In [13]:
# 1️⃣ Google Drive 마운트
from google.colab import drive
drive.mount('/content/drive')

# 2️⃣ 필요한 라이브러리 불러오기
import os
import pandas as pd

# 3️⃣ 데이터 폴더 경로 설정
data_folder = "/content/drive/MyDrive/1st_datathon"  # 폴더 경로

# 4️⃣ 폴더 내의 파일 목록 확인
file_list = os.listdir(data_folder)
print("📂 데이터 폴더 내 파일 목록:", file_list)

# 5️⃣ CSV 파일 경로 설정
disney_file = os.path.join(data_folder, "disney_plus_titles.csv")
amazon_file = os.path.join(data_folder, "amazon_prime_titles.csv")
netflix_file = os.path.join(data_folder, "netflix_titles.csv")

# 6️⃣ pandas로 CSV 파일 로드
disney_df = pd.read_csv(disney_file)
amazon_df = pd.read_csv(amazon_file)
netflix_df = pd.read_csv(netflix_file)

# 7️⃣ 컬럼 이름을 소문자로 통일
netflix_df.columns = netflix_df.columns.str.lower()
amazon_df.columns = amazon_df.columns.str.lower()
disney_df.columns = disney_df.columns.str.lower()

# 8️⃣ 공통 컬럼만 추출 (필요한 컬럼만 남기는 것으로 활용해도 OK)
common_cols = ['show_id', 'type', 'title', 'director', 'cast', 'release_year', 'date_added','listed_in','duration']
netflix_df = netflix_df[common_cols]
amazon_df = amazon_df[common_cols]
disney_df = disney_df[common_cols]

# 9️⃣ 출처 플랫폼 컬럼 추가
netflix_df['platform'] = 'Netflix'
amazon_df['platform'] = 'Amazon'
disney_df['platform'] = 'Disney'

# 🔟 데이터 크기 확인
print("✅ Disney+ 데이터:", disney_df.shape)
print("✅ Amazon Prime 데이터:", amazon_df.shape)
print("✅ Netflix 데이터:", netflix_df.shape)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
📂 데이터 폴더 내 파일 목록: ['disney_plus_titles.csv', 'amazon_prime_titles.csv', 'netflix_titles.csv']
✅ Disney+ 데이터: (1450, 10)
✅ Amazon Prime 데이터: (9668, 10)
✅ Netflix 데이터: (8807, 10)


<ipython-input-13-10468c4c6dc2>:40: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  disney_df['platform'] = 'Disney'


=> 각각 1450 / 9668 / 8807개의 항목(영화나 TV쇼 etc.)에 대한 12개의 정보(column) 존재 확인.

# 2. 모든 dataset에서 column 동일한지 확인

모든 OTT 데이터셋의 컬럼이 동일한지 검증 완료. ✅

In [14]:
# 각 데이터프레임의 컬럼 확인
print("🎬 Disney+ 컬럼:", disney_df.columns)
print("🎬 Amazon Prime 컬럼:", amazon_df.columns)
print("🎬 Netflix 컬럼:", netflix_df.columns)

# 모든 데이터셋에서 컬럼이 동일한지 확인
print("✅ 모든 데이터셋의 컬럼 동일 여부:",
      disney_df.columns.equals(amazon_df.columns) and disney_df.columns.equals(netflix_df.columns))

🎬 Disney+ 컬럼: Index(['show_id', 'type', 'title', 'director', 'cast', 'release_year',
       'date_added', 'listed_in', 'duration', 'platform'],
      dtype='object')
🎬 Amazon Prime 컬럼: Index(['show_id', 'type', 'title', 'director', 'cast', 'release_year',
       'date_added', 'listed_in', 'duration', 'platform'],
      dtype='object')
🎬 Netflix 컬럼: Index(['show_id', 'type', 'title', 'director', 'cast', 'release_year',
       'date_added', 'listed_in', 'duration', 'platform'],
      dtype='object')
✅ 모든 데이터셋의 컬럼 동일 여부: True


# 3. 결측치 확인

각 데이터셋의 결측치 개수 및 비율 확인 완료. ✅

In [15]:
# 데이터프레임별 결측치 개수 확인
print("🔍 Disney+ 결측치 개수:\n", disney_df.isnull().sum())
print("\n🔍 Amazon Prime 결측치 개수:\n", amazon_df.isnull().sum())
print("\n🔍 Netflix 결측치 개수:\n", netflix_df.isnull().sum())

# 전체 데이터 중 결측치 비율 확인
print("\n🔎 Disney+ 결측치 비율:\n", (disney_df.isnull().sum() / len(disney_df)) * 100)
print("\n🔎 Amazon Prime 결측치 비율:\n", (amazon_df.isnull().sum() / len(amazon_df)) * 100)
print("\n🔎 Netflix 결측치 비율:\n", (netflix_df.isnull().sum() / len(netflix_df)) * 100)

🔍 Disney+ 결측치 개수:
 show_id           0
type              0
title             0
director        473
cast            190
release_year      0
date_added        3
listed_in         0
duration          0
platform          0
dtype: int64

🔍 Amazon Prime 결측치 개수:
 show_id            0
type               0
title              0
director        2083
cast            1233
release_year       0
date_added      9513
listed_in          0
duration           0
platform           0
dtype: int64

🔍 Netflix 결측치 개수:
 show_id            0
type               0
title              0
director        2634
cast             825
release_year       0
date_added        10
listed_in          0
duration           3
platform           0
dtype: int64

🔎 Disney+ 결측치 비율:
 show_id          0.000000
type             0.000000
title            0.000000
director        32.620690
cast            13.103448
release_year     0.000000
date_added       0.206897
listed_in        0.000000
duration         0.000000
platform         0.0000

# 4. 중복data 확인

중복 데이터 없음 확인 완료. ✅

In [16]:
# 데이터프레임별 중복 데이터 개수 확인
print("📌 Disney+ 중복 데이터 개수:", disney_df.duplicated().sum())
print("📌 Amazon Prime 중복 데이터 개수:", amazon_df.duplicated().sum())
print("📌 Netflix 중복 데이터 개수:", netflix_df.duplicated().sum())

📌 Disney+ 중복 데이터 개수: 0
📌 Amazon Prime 중복 데이터 개수: 0
📌 Netflix 중복 데이터 개수: 0


# 5. 결측치 처리 및 data 병합

결측치 처리 및 데이터 병합 완료, 최종 데이터 19,925행 10컬럼 생성. ✅

In [17]:
# 1️⃣ 결측치 처리
for df in [disney_df, amazon_df, netflix_df]:
    df = df.copy()  # 경고 방지를 위해 복사본 사용

    # 'date_added' 컬럼이 존재하면 삭제
    if 'date_added' in df.columns:
        df = df.drop(columns=['date_added'])

    # 'director'와 'cast'의 결측치를 'Unknown'으로 채우기
    df['director'] = df['director'].fillna('Unknown')
    df['cast'] = df['cast'].fillna('Unknown')

    # 'duration'의 결측치가 3개뿐이므로 해당 행 제거
    df = df.dropna(subset=['duration'])

# 2️⃣ 데이터 병합
combined_df = pd.concat([disney_df, amazon_df, netflix_df], ignore_index=True)

# 확인
print("✅ 최종 데이터 크기:", combined_df.shape)
print(combined_df.head())

✅ 최종 데이터 크기: (19925, 10)
  show_id     type                                             title  \
0      s1    Movie  Duck the Halls: A Mickey Mouse Christmas Special   
1      s2    Movie                            Ernest Saves Christmas   
2      s3    Movie                      Ice Age: A Mammoth Christmas   
3      s4    Movie                        The Queen Family Singalong   
4      s5  TV Show                             The Beatles: Get Back   

                            director  \
0  Alonso Ramirez Ramos, Dave Wasson   
1                        John Cherry   
2                       Karen Disher   
3                    Hamish Hamilton   
4                                NaN   

                                                cast  release_year  \
0  Chris Diamantopoulos, Tony Anselmo, Tress MacN...          2016   
1           Jim Varney, Noelle Parker, Douglas Seale          1988   
2  Raymond Albert Romano, John Leguizamo, Denis L...          2011   
3  Darren Criss, Adam

# 6. 중복되는 영화/TV프로그램 제거

중복된 영화/TV 프로그램을 플랫폼 우선순위로 정리하여 중복 제거 완료, 최최종 데이터 8,804행 9컬럼. ✅

In [18]:
# 중복된 영화/TV 프로그램을 하나만 남기고 삭제
# 우선순위 설정: Netflix > Amazon Prime > Disney+
platform_order = ['Netflix', 'Amazon', 'Disney']

# 중복된 title에 대해 platform 기준으로 우선순위를 적용하여 남길 행을 결정
df_sorted = df.sort_values(by=['title', 'platform'], ascending=[True, True])
df_sorted['platform'] = pd.Categorical(df_sorted['platform'], categories=platform_order, ordered=True)

# 중복된 title을 가진 행에서 platform 순서대로 중복 제거
df_deduplicated = df_sorted.drop_duplicates(subset=['title'], keep='first')

# 결과 확인
print(f'중복 제거 후 데이터 크기: {df_deduplicated.shape}')

중복 제거 후 데이터 크기: (8804, 9)


# 7. 장르 파악 (listed_in)

'listed_in'에서 장르를 분리하고 출현 빈도 계산 완료, 상위 20개 장르 출력. ✅

In [19]:
# 1단계: listed_in 컬럼에서 장르 목록을 추출하여 장르의 출현 빈도 계산하기
# 각 영화/TV 프로그램의 listed_in이 여러 개의 장르를 포함할 수 있으므로, 먼저 이를 분리하고 나열된 장르들을 확인

# 'listed_in' 컬럼에서 장르를 분리하여 나열
all_genres = df_deduplicated['listed_in'].str.split(',').explode().str.strip()

# 장르별 출현 빈도 계산
genre_counts = all_genres.value_counts()

# 결과 확인
print(genre_counts.head(20))  # 상위 20개의 장르를 출력

listed_in
International Movies        2752
Dramas                      2427
Comedies                    1674
International TV Shows      1351
Documentaries                869
Action & Adventure           859
TV Dramas                    763
Independent Movies           756
Children & Family Movies     641
Romantic Movies              616
TV Comedies                  581
Thrillers                    577
Crime TV Shows               470
Kids' TV                     451
Docuseries                   395
Music & Musicals             375
Romantic TV Shows            370
Horror Movies                357
Stand-Up Comedy              343
Reality TV                   255
Name: count, dtype: int64


# 8. 전체 장르 고유값 파악

'listed_in'에서 고유한 장르 목록 추출 완료. ✅

In [20]:
# 1. listed_in 컬럼에 등장하는 고유한 값들을 확인
unique_genres = df_deduplicated['listed_in'].unique()

# 2. 고유한 장르 목록 출력
print(unique_genres)

['Horror Movies, International Movies, Thrillers'
 'Documentaries, International Movies'
 'Dramas, International Movies, Romantic Movies'
 'Comedies, Romantic Movies' 'Documentaries, Sports Movies'
 'Comedies, Dramas, International Movies' 'TV Comedies' 'Comedies'
 'Dramas, International Movies' 'Sports Movies' 'Documentaries'
 'Reality TV' 'Dramas, Romantic Movies'
 'Dramas, Romantic Movies, Sports Movies'
 'Comedies, International Movies, Romantic Movies' 'Action & Adventure'
 'International TV Shows, Romantic TV Shows, TV Dramas'
 'Docuseries, Science & Nature TV'
 'Dramas, International Movies, Sports Movies' 'Movies'
 'British TV Shows, International TV Shows, Reality TV'
 'Dramas, Independent Movies, Sports Movies'
 'International TV Shows, Korean TV Shows, Romantic TV Shows'
 'Horror Movies, International Movies'
 'Horror Movies, Independent Movies, Thrillers'
 'Crime TV Shows, TV Dramas, TV Mysteries' 'Crime TV Shows, Docuseries'
 'Horror Movies, Thrillers' 'Action & Adventure,

# 9. 장르 개수 파악 및 대분류 카테고리화

'listed_in'에서 세부 장르를 분리하여 유니크한 장르 42개 추출 완료. ✅

In [21]:
import re

# 작은따옴표 제거 후, ','로 세부 장르 분리하기
cleaned_genres = []

for genre in unique_genres:
    # 작은따옴표 제거
    genre = genre.replace("'", "")
    # ','로 세부 장르 분리
    cleaned_genres.extend(genre.split(', '))

# 중복 제거 후 유니크한 세부 장르 세기
unique_cleaned_genres = set(cleaned_genres)
num_unique_genres = len(unique_cleaned_genres)


print(f"유니크한 세부 장르의 개수: {num_unique_genres}")
print("유니크한 세부 장르 목록:")
for genre in unique_cleaned_genres:
    print(genre)

유니크한 세부 장르의 개수: 42
유니크한 세부 장르 목록:
Cult Movies
Classic Movies
Stand-Up Comedy
Dramas
Kids TV
Sci-Fi & Fantasy
Teen TV Shows
British TV Shows
LGBTQ Movies
TV Horror
Action & Adventure
Horror Movies
TV Action & Adventure
Classic & Cult TV
Crime TV Shows
Music & Musicals
TV Dramas
Thrillers
Faith & Spirituality
Spanish-Language TV Shows
International TV Shows
International Movies
Children & Family Movies
TV Shows
Anime Series
Comedies
Korean TV Shows
TV Comedies
Romantic TV Shows
Science & Nature TV
Sports Movies
Stand-Up Comedy & Talk Shows
Documentaries
Docuseries
TV Mysteries
Independent Movies
TV Thrillers
TV Sci-Fi & Fantasy
Movies
Anime Features
Romantic Movies
Reality TV


# 10. 10개 대분류 카테고리와 해당 세부 장르 매핑

10개 대분류 카테고리와 세부 장르 매핑 완료. ✅

In [39]:
genre_mapping = {
    'Action': [
        'Action & Adventure', 'TV Action & Adventure'  # 2개
    ],
    'Crime': [
        'Crime TV Shows', 'TV Mysteries'  # 3개
    ],
    'Sci-Fi/Fantasy': [
        'Sci-Fi & Fantasy', 'TV Sci-Fi & Fantasy', 'Anime Features', 'Anime Series', 'TV Sci-Fi & Fantasy'  # 5개
    ],
    'Comedy': [
        'Stand-Up Comedy', 'TV Comedies', 'Comedies', 'Stand-Up Comedy & Talk Shows'  # 4개
    ],
    'Horror/Thriller': [
        'TV Horror', 'Horror Movies', 'TV Thrillers', 'Thrillers', 'TV Action & Adventure', 'Cult Movies'  # 6개
    ],
    'War': [
        'War Movies'  # 1개
    ],
    'Sports': [
        'Sports Movies', 'Reality TV'  # 2개
    ],
    'Music/Musical': [
        'Music & Musicals', 'Stand-Up Comedy & Talk Shows'  # 2개
    ],
    'Romance/Love': [
        'Romantic TV Shows', 'Romantic Movies'  # 2개
    ],
    'Classic': [
        'Documentaries', 'Docuseries', 'Science & Nature TV', 'Independent Movies'  # 3개
    ]
}

# 각 대분류 장르별 소분류 개수 확인
total_subgenres = 0
for category, subgenres in genre_mapping.items():
    print(f"{category}: {len(subgenres)} 개")

    total_subgenres += len(subgenres)

print(f"\n전체 소분류 장르 개수: {total_subgenres} 개")

Action: 5 개
Crime: 3 개
Sci-Fi/Fantasy: 5 개
Comedy: 4 개
Horror/Thriller: 5 개
War: 1 개
Sports: 2 개
Music/Musical: 2 개
Romance/Love: 2 개
Documentary: 3 개

전체 소분류 장르 개수: 32 개


# 11. 대분류 카테고리화

listed_in 컬럼을 기반으로 세부 장르를 대분류 카테고리로 매핑하여 새로운 category 컬럼을 추가 ✅

In [28]:
# 4️⃣ 세부 장르를 대분류 카테고리로 변환하는 함수
def map_genres_to_categories(listed_in, genre_mapping):
    # 작은따옴표 제거
    listed_in = listed_in.replace("'", "")

    # 결과로 대분류 카테고리 제목만 담을 리스트
    categories = []

    # 세부 장르들을 대분류 카테고리로 매핑
    for category, genres in genre_mapping.items():
        for genre in genres:
            if genre in listed_in:
                categories.append(category)
                break  # 하나의 대분류 카테고리만 추가 (중복 방지)

    # 대분류 카테고리들이 없으면 빈 문자열 반환
    return ', '.join(categories) if categories else ''

# 5️⃣ 'listed_in' 컬럼을 대분류 카테고리로 변환
df_deduplicated['category'] = df_deduplicated['listed_in'].apply(lambda x: map_genres_to_categories(x, genre_mapping))

# 6️⃣ 결과 확인
print(df_deduplicated[['title', 'platform', 'listed_in', 'category']])

                              title platform  \
2036                         #Alive  Netflix   
2304  #AnneFrank - Parallel Stories  Netflix   
2481              #FriendButMarried  Netflix   
2324            #FriendButMarried 2  Netflix   
5973                          #Roxy  Netflix   
...                             ...      ...   
6177                       忍者ハットリくん  Netflix   
4914                           海的儿子  Netflix   
7101                           마녀사냥  Netflix   
5022                        반드시 잡는다  Netflix   
7108            최강전사 미니특공대 : 영웅의 탄생  Netflix   

                                              listed_in  \
2036     Horror Movies, International Movies, Thrillers   
2304                Documentaries, International Movies   
2481      Dramas, International Movies, Romantic Movies   
2324      Dramas, International Movies, Romantic Movies   
5973                          Comedies, Romantic Movies   
...                                                 ...   
6177      

# 12. 한 영화/TV당 한 장르 배정

각 작품에 대해 가장 우선순위가 높은 카테고리를 top_category 컬럼으로 배정 완료. ✅

In [29]:
import pandas as pd
from collections import Counter
from IPython.display import display

# df_deduplicated를 기반으로 작업
# 각 작품의 category를 쉼표로 분할하여 리스트로 변환
df_deduplicated["category_list"] = df_deduplicated["listed_in"].apply(lambda x: x.split(", "))

# 모든 category를 모아서 빈도수 계산
all_categories = [cat for sublist in df_deduplicated["category_list"] for cat in sublist]
category_counts = Counter(all_categories)

# 카테고리를 빈도수 기준으로 정렬
top_categories = sorted(category_counts.items(), key=lambda x: x[1], reverse=True)
print("카테고리 빈도수 순위:", top_categories)

# 각 작품에 대해 가장 우선순위가 높은 대분류 배정
def get_top_category(categories):
    sorted_categories = sorted(categories, key=lambda x: category_counts[x], reverse=True)
    return sorted_categories[0] if sorted_categories else None

df_deduplicated["top_category"] = df_deduplicated["category_list"].apply(get_top_category)

# 결과 출력
print(df_deduplicated[["title", "top_category"]])

display(df_deduplicated.head(20))

카테고리 빈도수 순위: [('International Movies', 2752), ('Dramas', 2427), ('Comedies', 1674), ('International TV Shows', 1351), ('Documentaries', 869), ('Action & Adventure', 859), ('TV Dramas', 763), ('Independent Movies', 756), ('Children & Family Movies', 641), ('Romantic Movies', 616), ('TV Comedies', 581), ('Thrillers', 577), ('Crime TV Shows', 470), ("Kids' TV", 451), ('Docuseries', 395), ('Music & Musicals', 375), ('Romantic TV Shows', 370), ('Horror Movies', 357), ('Stand-Up Comedy', 343), ('Reality TV', 255), ('British TV Shows', 253), ('Sci-Fi & Fantasy', 243), ('Sports Movies', 219), ('Anime Series', 176), ('Spanish-Language TV Shows', 174), ('TV Action & Adventure', 168), ('Korean TV Shows', 151), ('Classic Movies', 116), ('LGBTQ Movies', 102), ('TV Mysteries', 98), ('Science & Nature TV', 92), ('TV Sci-Fi & Fantasy', 84), ('TV Horror', 75), ('Cult Movies', 71), ('Anime Features', 71), ('Teen TV Shows', 69), ('Faith & Spirituality', 65), ('TV Thrillers', 57), ('Stand-Up Comedy & Talk

show_id     type                          title  \
2036   s2037    Movie                         #Alive   
2304   s2305    Movie  #AnneFrank - Parallel Stories   
2481   s2482    Movie              #FriendButMarried   
2324   s2325    Movie            #FriendButMarried 2   
5973   s5974    Movie                          #Roxy   
5695   s5696    Movie                      #Rucker50   
654     s655    Movie                        #Selfie   
655     s656    Movie                     #Selfie 69   
2666   s2667  TV Show                       #blackAF   
5972   s5973    Movie               #cats_the_mewvie   
5277   s5278    Movie                   #realityhigh   
306     s307    Movie                            '76   
5967   s5968    Movie                            '89   
5971   s5972    Movie                       (T)ERROR   
2140   s2141  TV Show                       (Un)Well   
5979   s5980    Movie               1 Chance 2 Dance   
5980   s5981    Movie                  1 Mile to You   
3392   s3393    Movie            10 Days in Sun City   
5387   s5388    Movie                 10 jours en or   
5981   s5982    Movie                    10,000 B.C.   

                                    director  \
2036                                  Cho Il   
2304             Sabina Fedeli, Anna Migotto   
2481                           Rako Prijanto   
2324                           Rako Prijanto   
5973                         Michael Kennedy   
5695                   Robert McCullough Jr.   
654                           Cristina Jacob   
655                           Cristina Jacob   
2666                                 Unknown   
5972                        Michael Margolis   
5277                        Fernando Lebrija   
306                               Izu Ojukwu   
5967                                 Unknown   
5971  Lyric R. Cabral, David Felix Sutcliffe   
2140                                 Unknown   
5979                              Adam Deyoe   
5980                             Leif Tilden   
3392                               Adze Ugah   
5387                       Nicolas Brossette   
5981                         Roland Emmerich   

                                                   cast  release_year  \
2036                           Yoo Ah-in, Park Shin-hye          2020   
2304                        Helen Mirren, Gengher Gatti          2019   
2481  Adipati Dolken, Vanesha Prescilla, Rendi Jhon,...          2018   
2324  Adipati Dolken, Mawar de Jongh, Sari Nila, Von...          2020   
5973  Jake Short, Sarah Fisher, Booboo Stewart, Dann...          2018   
5695                                            Unknown          2016   
654   Flavia Hojda, Crina Semciuc, Olimpia Melinte, ...          2014   
655   Maia Morgenstern, Olimpia Melinte, Crina Semci...          2016   
2666  Kenya Barris, Rashida Jones, Iman Benson, Genn...          2020   
5972                                            Unknown          2020   
5277  Nesta Cooper, Kate Walsh, John Michael Higgins...          2017   
306   Ramsey Nouah, Rita Dominic, Chidi Mokeme, Ibin...          2016   
5967                 Lee Dixon, Ian Wright, Paul Merson          2017   
5971                                            Unknown          2015   
2140                                            Unknown          2020   
5979  Lexi Giovagnoli, Justin Ray, Rae Latt, Poonam ...          2014   
5980  Billy Crudup, Graham Rogers, Liana Liberato, S...          2017   
3392  Ayo Makun, Adesua Etomi, Richard Mofe-Damijo, ...          2017   
5387  Franck Dubosc, Claude Rich, Marie Kremer, Math...          2012   
5981  Steven Strait, Camilla Belle, Cliff Curtis, Jo...          2008   

                                            listed_in  duration platform  \
2036   Horror Movies, International Movies, Thrillers    99 min  Netflix   
2304              Documentaries, International Movies    95 min  Netflix   
2481    Dramas, International Movies, Romantic Movies   1

# 13. cast 칼럼을 actor 칼럼으로 쪼개기

In [36]:
import pandas as pd
from IPython.display import display

# 3️⃣ 배우 데이터 정리: 'cast' 컬럼에서 배우별로 분리하여 개별 행으로 확장

# df_deduplicated_2를 df_deduplicated의 복사본으로 생성
df_deduplicated_2 = df_deduplicated.copy()

# 'cast' 컬럼에서 쉼표로 구분된 배우들을 분리하여 리스트로 만듦
df_deduplicated_2['actor'] = df_deduplicated_2['cast'].apply(lambda x: x.split(', ') if isinstance(x, str) else [])

# 'actor'를 기준으로 explode()하여 각 배우가 한 행을 차지하도록 만듦
df_deduplicated_2 = df_deduplicated_2.explode('actor').reset_index(drop=False)

# 'actor' 컬럼이 생성된 후 'cast'는 더 이상 필요 없으므로 삭제
df_deduplicated_2 = df_deduplicated_2.drop(columns=['cast'])

# 결과를 표 형식으로 보기 좋게 출력
# 'index'를 포함시키기 위해 reset_index(drop=False)로 변경
display(df_deduplicated_2[['index', 'show_id', 'type', 'title', 'director', 'actor', 'release_year', 'listed_in', 'duration', 'platform', 'category', 'category_list', 'top_category']].tail(20))

index show_id     type                 title       director  \
64928   4667   s4668    Movie  ​​Kuch Bheege Alfaaz           Onir   
64929   4667   s4668    Movie  ​​Kuch Bheege Alfaaz           Onir   
64930   6177   s6178  TV Show              忍者ハットリくん        Unknown   
64931   4914   s4915  TV Show                  海的儿子        Unknown   
64932   4914   s4915  TV Show                  海的儿子        Unknown   
64933   4914   s4915  TV Show                  海的儿子        Unknown   
64934   4914   s4915  TV Show                  海的儿子        Unknown   
64935   4914   s4915  TV Show                  海的儿子        Unknown   
64936   7101   s7102  TV Show                  마녀사냥        Unknown   
64937   7101   s7102  TV Show                  마녀사냥        Unknown   
64938   7101   s7102  TV Show                  마녀사냥        Unknown   
64939   7101   s7102  TV Show                  마녀사냥        Unknown   
64940   7101   s7102  TV Show                  마녀사냥        Unknown   
64941   5022   s5023    Movie               반드시 잡는다  Hong-seon Kim   
64942   7108   s7109    Movie   최강전사 미니특공대 : 영웅의 탄생  Young Jun Lee   
64943   7108   s7109    Movie   최강전사 미니특공대 : 영웅의 탄생  Young Jun Lee   
64944   7108   s7109    Movie   최강전사 미니특공대 : 영웅의 탄생  Young Jun Lee   
64945   7108   s7109    Movie   최강전사 미니특공대 : 영웅의 탄생  Young Jun Lee   
64946   7108   s7109    Movie   최강전사 미니특공대 : 영웅의 탄생  Young Jun Lee   
64947   7108   s7109    Movie   최강전사 미니특공대 : 영웅의 탄생  Young Jun Lee   

                     actor  release_year  \
64928  Saheb Bhattacharjee          2018   
64929      Shefali Chauhan          2018   
64930              Unknown          2012   
64931           Li Nanxing          2016   
64932      Christopher Lee          2016   
64933          Jesseca Liu          2016   
64934           Apple Hong          2016   
64935          Jeanette Aw          2016   
64936        Si-kyung Sung          2015   
64937          Se-yoon Yoo          2015   
64938        Dong-yup Shin          2015   
64939         Ji-woong Heo          2015   
64940       Sam Hammington          2015   
64941        Baek Yoon-sik          2017   
64942         Um Sang-hyun          2018   
64943       Yang Jeong-hwa          2018   
64944        Jeon Tae-yeol          2018   
64945        Shin Yong-woo          2018   
64946         Lee So-young          2018   
64947              So-yeon          2018   

                                               listed_in   duration platform  \
64928   Dramas, Independent Movies, International Movies    110 min  Netflix   
64929   Dramas, Independent Movies, International Movies    110 min  Netflix   
64930                             Anime Series, Kids' TV  2 Seasons  Netflix   
64931                  International TV Shows, TV Dramas   1 Season  Netflix   
64932                  International TV Shows, TV Dramas   1 Season  Netflix   
64933                  International TV Shows, TV Dramas   1 Season  Netflix   
64934                  International TV Shows, TV Dramas   1 Season  Netflix   
64935                  International TV Shows, TV Dramas   1 Season  Netflix   
64936  International TV Shows, Korean TV Shows, Stand...   1 Season  Netflix   
64937  International TV Shows, Korean TV Shows, Stand...   1 Season  Netflix   
64938  International TV Shows, Korean TV Shows, Stand...   1 Season  Netflix   
64939  International TV Shows, Korean TV Shows, Stand...   1 Season  Netflix   
64940  International TV Shows, Korean TV Shows, Stand...   1 Season  Netflix   
64941            Dramas, International Movies, Thrillers    110 min  Netflix   
64942                           Children & Family Movies     68 min  Netflix   
64943                           Children & Family Movies     68 min  Netflix   
64944                           Children & Family Movies     68 min  Netflix   
64945                           Children & Family Movies     68 min  Netflix   
64946                           Children & Family Movies     68 min  Netflix   
64947      

# 14.

In [35]:
# 필요한 라이브러리 불러오기
import pandas as pd

# top_category 별로 배우 출연 횟수 계산
actor_counts_by_genre = df_deduplicated_2.groupby(['top_category', 'actor']).size().reset_index(name='actor_count')

# 각 장르별로 배우 출연 횟수를 내림차순으로 정렬
actor_counts_by_genre_sorted = actor_counts_by_genre.sort_values(['top_category', 'actor_count'], ascending=[True, False])

# 각 장르별로 상위 N명 배우를 출력 (여기서는 상위 5명을 예시로 출력)
top_n = 5  # 상위 N명의 배우를 선택할 수 있음
top_actors_per_genre = actor_counts_by_genre_sorted.groupby('top_category').head(top_n)

# 결과를 보기 좋게 출력
# 각 장르별로 상위 배우들을 출력
for genre in top_actors_per_genre['top_category'].unique():
    print(f"\n장르: {genre}")
    display(top_actors_per_genre[top_actors_per_genre['top_category'] == genre][['actor', 'actor_count']])


장르: Action & Adventure


actor  actor_count
269    Bruce Willis            9
1527   Nicolas Cage            8
466     Danny Trejo            7
675    Frank Grillo            7
762   Harrison Ford            7


장르: Anime Features


actor  actor_count
2199    Anne Watanabe            1
2200  Barbara Goodson            1
2201   Cindy Robinson            1
2202   Erica Lindbeck            1
2203     Ethan Murray            1


장르: Anime Series


actor  actor_count
2231     Bill Rogers            3
2250    Edward Bosco            3
2256    Frank Todaro            3
2273    Jake Foushee            3
2275  Jason Marnocha            3


장르: British TV Shows


actor  actor_count
2364          Andy Nyman            1
2365      Beth Cordingly            1
2366       Davina McCall            1
2367      Jaime Winstone            1
2368  Kathleen McDermott            1


장르: Children & Family Movies


actor  actor_count
3042   Julie Tejwani           26
3490    Rupa Bhimani           25
3410     Rajesh Kava           21
2966  Jigna Bhardwaj           18
3661         Unknown           18


장르: Comedies


actor  actor_count
3753  Adam Sandler           17
6481  Maya Rudolph           14
5663   John Cleese           12
7835  Will Ferrell           12
4668   David Spade           11


장르: Crime TV Shows


actor  actor_count
8147           Unknown           50
7985  Deborah Ann Woll            2
8036      Jon Bernthal            2
8110      Piers Morgan            2
8113          R. Kelly            2


장르: Documentaries


actor  actor_count
8884             Unknown          262
8320       Craig Sechler            6
8731        Quincy Jones            5
8341  David Attenborough            4
8490      Jay O. Sanders            4


장르: Docuseries


actor  actor_count
9183             Unknown           92
8976  David Attenborough            9
8978         David Chang            3
9113      Morgan Freeman            3
8936    Anthony Bourdain            2


장르: Dramas


actor  actor_count
11662      James Franco           11
12576   Kristen Stewart           11
14262  Robert Pattinson           11
9500          Amy Adams           10
14247    Robert De Niro            9


장르: Horror Movies


actor  actor_count
15545      Ashley Zukerman            4
16056            Tony Todd            4
15553  Benjamin Flores Jr.            3
15741        James Ransone            3
15787        Jordana Spiro            3


장르: Independent Movies


actor  actor_count
16155       Carmen Ejogo            2
16219       Emile Hirsch            2
16224       Enver Gjokaj            2
16228         Flora Diaz            2
16250  Heidi Niedermeyer            2


장르: International Movies


actor  actor_count
29000           Unknown          178
17748       Anupam Kher           38
27533    Shah Rukh Khan           32
16966      Akshay Kumar           27
25000  Naseeruddin Shah           27


장르: International TV Shows


actor  actor_count
38019           Unknown          109
37635  Takahiro Sakurai           22
38532         Yuki Kaji           18
30150         Ai Kayano           16
33715    Junichi Suwabe           16


장르: Kids' TV


actor  actor_count
40391            Unknown           30
40409       Vincent Tong           13
38902      Ashleigh Ball            9
39911       Michela Luci            9
39551  Johnny Yong Bosch            7


장르: Movies


actor  actor_count
40658       Stephen Fry            5
40494  Cassandra Morris            4
40668           Unknown            4
40466        Abby Trott            3
40470     Alex Heartman            3


장르: Music & Musicals


actor  actor_count
40720  Minako Kotobuki            2
40742      Tim Minchin            2
40676     Aki Toyosaki            1
40677      Akira Emoto            1
40678   Alaina Huffman            1


장르: Reality TV


actor  actor_count
40930         Unknown           38
40879  Paul Hollywood            4
40897   Rutledge Wood            4
40823    Hunter March            3
40864      Mary Berry            3


장르: Romantic Movies


actor  actor_count
40938   Allen Maldonado            1
40939   Allison Hossack            1
40940        Bea Santos            1
40941  Bradley Hamilton            1
40942       Brely Evans            1


장르: Romantic TV Shows


actor  actor_count
41048           Unknown            7
41011     Kana Hanazawa            2
41027     Mamoru Miyano            2
41038  Nobuhiko Okamoto            2
40987         Ai Kayano            1


장르: Sci-Fi & Fantasy


actor  actor_count
41056          David Tse            1
41057  Joe David Walters            1
41058          Julie Cox            1
41059     Katee Sackhoff            1
41060         Ray Fearon            1


장르: Sports Movies


actor  actor_count
41062   Ian Wright            1
41063    Lee Dixon            1
41064  Paul Merson            1


장르: Stand-Up Comedy


actor  actor_count
41194       Jeff Dunham            8
41232        Kevin Hart            8
41226     Katt Williams            6
41098         Bill Burr            5
41185  Iliza Shlesinger            5


장르: Stand-Up Comedy & Talk Shows


actor  actor_count
41389          Unknown            2
41374      Ari Shaffir            1
41375         Bill Nye            1
41376  David Letterman            1
41377     Derek Muller            1


장르: TV Action & Adventure


actor  actor_count
41507  Kevin Michael Richardson            2
41538              Michael Dorn            2
41390          Aaron Himelstein            1
41391              Adam Bartley            1
41392             Adam Tsekhman            1


장르: TV Comedies


actor  actor_count
43172           Unknown           16
42107  Fortune Feimster            5
42171      Grey Griffin            5
43154         Tony Hale            5
41945       David Cross            4


장르: TV Dramas


actor  actor_count
44901         Mike Colter            5
44054  Giancarlo Esposito            4
44467          Kate Walsh            4
43473          Ben Lawson            3
43595      Campbell Scott            3


장르: TV Horror


actor  actor_count
45645            Alex Lange            1
45646        Aleyse Shannon            1
45647           An-Li Bogan            1
45648     Christopher Shyer            1
45649  David Clayton Rogers            1


장르: TV Mysteries


actor  actor_count
45688  Caleb McLaughlin            2
45693    Charlie Heaton            2
45700  Dacre Montgomery            2
45703     David Harbour            2
45718     Finn Wolfhard            2


장르: TV Sci-Fi & Fantasy


actor  actor_count
45840  Alex Mallari Jr.            1
45841     Anthony Lemke            1
45842   Jodelle Ferland            1
45843     Marc Bendavid            1
45844    Melissa O'Neil            1


장르: TV Shows


actor  actor_count
45936         Unknown            5
45862        Bob Ross            2
45848  Abeer Alotaibi            1
45849      Aditi Arya            1
45850     Aditi Balan            1


장르: Thrillers


actor  actor_count
46789    Nicolas Cage            4
46820  Patrick Wilson            4
45985      Ali Larter            3
46065    Ben Kingsley            3
46118    Cam Gigandet            3